crypto/bindings/conversion.ts

```diff
@@ -98,3 +158,120 @@ function fieldsFromRustFlat(fieldBytes: Uint8Array): MlArray<Field> {
   }
   return [0, ...fields];
 }
+
+// affine
+
+type WasmAffine = wasmNamespace.WasmGVesta | wasmNamespace.WasmGPallas;
+type MakeAffine<A extends WasmAffine> = () => A & { ptr: number };
+
+function affineFromRust(pt: WasmAffine): OrInfinity {
+  if (pt.infinity) {
+    pt.free();
+    return 0;
+  } else {
+    let x = fieldFromRust(pt.x);
+    let y = fieldFromRust(pt.y);
+    pt.free();
+    return [0, [0, x, y]];
+  }
+}
+
+function affineToRust<A extends WasmAffine>(
+  pt: OrInfinity,
+  klass: MakeAffine<A>
+) {
+  var res = klass();
+  if (pt === 0) {
+    res.infinity = true;
+  } else {
+    res.x = fieldToRust(pt[1][1]);
+    res.y = fieldToRust(pt[1][2]);
+  }
+  return res;
+}
+
+// polycomm
+
+type WasmPolyComm = wasmNamespace.WasmFpPolyComm | wasmNamespace.WasmFqPolyComm;
+type WasmPolyCommClass = wasm['WasmFpPolyComm'] | wasm['WasmFqPolyComm'];
+
+function polyCommFromRust(
+  polyComm: WasmPolyComm,
+  klass: WrapperClass<WasmAffine>,
+  shouldFree: boolean
+): PolyComm {
+  let rustShifted = polyComm.shifted;
+  let rustUnshifted = polyComm.unshifted;
+  let mlShifted: MlOption<OrInfinity> =
+    rustShifted === undefined ? 0 : [0, affineFromRust(rustShifted)];
+  let mlUnshifted = mlArrayFromRustVector(
+    rustUnshifted,
+    klass,
+    affineFromRust,
+    shouldFree
+  );
+  return [0, mlUnshifted, mlShifted];
+}
+
+function polyCommToRust(
+  [, camlUnshifted, camlShifted]: PolyComm,
+  PolyComm: WasmPolyCommClass,
+  makeAffine: MakeAffine<WasmAffine>
+): WasmPolyComm {
+  let rustShifted =
+    camlShifted === 0 ? undefined : affineToRust(camlShifted[1], makeAffine);
+  let rustUnshifted = mlArrayToRustVector(
+    camlUnshifted,
+    affineToRust,
+    makeAffine
+  );
+  return new PolyComm(rustUnshifted, rustShifted);
+}
+
+// generic rust helpers
+
+type Freeable = { free(): void };
+type WrappedPointer = Freeable & { ptr: number };
+type WrapperClass<T extends Freeable> = {
+  __wrap(i: number): T;
+};
+
+const registry = new FinalizationRegistry((ptr: WrappedPointer) => {
+  ptr.free();
+});
+
+function mlArrayFromRustVector<TRust extends Freeable, TMl>(
+  rustVector: Uint32Array,
+  klass: WrapperClass<TRust>,
+  convert: (c: TRust) => TMl,
+  shouldFree: boolean
+): MlArray<TMl> {
+  var n = rustVector.length;
+  var array: TMl[] = new Array(n);
+  for (let i = 0; i < n; i++) {
+    var rustValue = klass.__wrap(rustVector[i]);
+    array[i] = convert(rustValue);
+    if (shouldFree) rustValue.free();
+  }
+  return [0, ...array];
```

Comment:
Is it copying a new array or appending at the beginning? An optim would be to avoid this duplicate allocation and instantiate directly `Array(n + 1)`.

Reply:
Yes we could slightly optimize the code, but it comes at the cost that we have to always deal with the index translation by 1. This has led to bugs in the previous code base, so I'd like to keep it simple for now and always to the conversion to/from an ml array as a separate step.
In theory we could use `array.unshift(0)` at the end to get the same result without copying the array. The only reason I don't do that is because TS doesn't understand it

Reply:
Could we keep track in an issue of this kind of optimisations we could do? Just for the future us :).

Reply:
Opened follow-up issue: https://github.com/o1-labs/o1js-bindings/issues/154